File: services/order-service/src/main/java/dev/junyoung/exchange/orderservice/domain/model/entity/Trade.java
package dev.junyoung.exchange.orderservice.domain.model.entity;

import dev.junyoung.exchange.orderservice.domain.exception.OrderInvalidException;
import dev.junyoung.exchange.orderservice.domain.model.enums.Side;
import dev.junyoung.exchange.orderservice.domain.model.value.*;

import java.time.Instant;

public record Trade (
	TradeId tradeId,
	Symbol symbol,
	OrderId orderId,
	OrderId matchOrderId,
	Side side,
	Price price,
	Quantity quantity,
	QuoteQty quoteQty,
	Instant tradeAt,
	Instant createdAt
) {
	public Trade {
		if (tradeId == null) throw new OrderInvalidException("체결 ID는 필수입니다.");
		if (symbol == null) throw new OrderInvalidException("거래 심볼은 필수입니다.");
		if (orderId == null) throw new OrderInvalidException("주문 ID는 필수입니다.");
		if (matchOrderId == null) throw new OrderInvalidException("상대 주문 ID는 필수입니다.");
		if (side == null) throw new OrderInvalidException("주문 방향(매수/매도)은 필수입니다.");
		if (price == null) throw new OrderInvalidException("체결 가격은 필수입니다.");
		if (quantity == null) throw new OrderInvalidException("체결 수량은 필수입니다.");
		if (quantity.isZero()) throw new OrderInvalidException("체결 수량은 0보다 커야합니다.");
		if (quoteQty == null) throw new OrderInvalidException("체결 금액은 필수입니다.");
		if (quoteQty.isZero()) throw new OrderInvalidException("체결 금액은 0보다 커야합니다.");
		if (tradeAt == null) throw new OrderInvalidException("체결 시점은 필수입니다.");
	}

	public static Trade buyOf(
		TradeId tradeId,
		Symbol symbol,
		OrderId orderId,
		OrderId matchOrderId,
		Price price,
		Quantity quantity,
		QuoteQty quoteQty,
		Instant tradeAt
	) {
		return new Trade(
			tradeId,
			symbol,
			orderId,
			matchOrderId,
			Side.BUY,
			price,
			quantity,
			quoteQty,
			tradeAt,
			Instant.now()
		);
	}

	public static Trade sellOf(
		TradeId tradeId,
		Symbol symbol,
		OrderId orderId,
		OrderId matchOrderId,
		Price price,
		Quantity quantity,
		QuoteQty quoteQty,
		Instant tradeAt
	) {
		return new Trade(
			tradeId,
			symbol,
			orderId,
			matchOrderId,
			Side.SELL,
			price,
			quantity,
			quoteQty,
			tradeAt,
			Instant.now()
		);
	}
}

File: services/matching-engine/src/models/trade.rs
use crate::models::{AccountId, OrderId, Price, Quantity, QuoteQty, Side, Symbol, TradeId};
use std::time::Instant;

pub struct TradeParticipant {
    pub account_id: AccountId,
    pub order_id: OrderId,
}

pub struct Trade {
    pub trade_id: TradeId,
    pub symbol: Symbol,
    pub buy_account_id: AccountId,
    pub buy_order_id: OrderId,
    pub sell_account_id: AccountId,
    pub sell_order_id: OrderId,
    pub price: Price,
    pub quantity: Quantity,
    pub quote_qty: QuoteQty,
    pub trade_at: Instant,
}

impl Trade {
    pub fn new(
        symbol: Symbol,
        taker_side: Side,
        taker: TradeParticipant,
        maker: TradeParticipant,
        price: Price,
        quantity: Quantity,
        quote_qty: QuoteQty,
    ) -> Self {
        let (buy, sell) = match taker_side {
            Side::Buy => (taker, maker),
            Side::Sell => (maker, taker),
        };

        Self {
            trade_id: TradeId::new(),
            symbol,
            buy_account_id: buy.account_id,
            buy_order_id: buy.order_id,
            sell_account_id: sell.account_id,
            sell_order_id: sell.order_id,
            price,
            quantity,
            quote_qty,
            trade_at: Instant::now(),
        }
    }
}
